### PR TITLE
Feature/verbose reporting

### LINKED ISSUE #12: Report number of total tests and number of passed/failed test.
TSD extracts the assertions in a test file. We can count the number of assertions and then return it back so further data would be made available for diagnosis. It will also be helpful when integrating TSD with test runners like Jest or AVA.

**Thoughts**

The `numTests`, `numFailedTests` and `numPassedTests` should be made available only if the TSD library is invoked with `verbose: true` option. Something like this,

```js
const diagnose = await tsd({
  cwd: path.join(__dirname, '/custom/path'),
  verbose: true
});
```

### PR DESCRIPTION
**Pull request by smartclash**
> This PR is stacked upon #9.

TSD either returns a null array if there is no test failures or an array of [`Diagnostic`](https://github.com/MLH-Fellowship/tsd/blob/8a9c90ee7f6e25e2108a87f5ac85338237404201/source/lib/interfaces.ts#L33-L39) objects. This PR aims to make TSD return with [`ExtendedDiagnostic`](https://github.com/MLH-Fellowship/tsd/blob/8a9c90ee7f6e25e2108a87f5ac85338237404201/source/lib/interfaces.ts#L41-L44) object no matter if the tests fail or not. 

Merging this PR allows custom test runners like Jest to report the number of assertions it evaluated in each file.

It also means that it introduces a breaking change because TSD now returns...

- An object with `numTests` and null `diagnostics` array if all tests passed.
```js
{
    numTests: 1024, // Total number of tests it ran through
    diagnostics: [] // No failed test cases
}
```

- An object with `numTests` and a `diagnostics` array populated with test fail reports if a test is failed.
```js
{
    numTests: 1024, // Total number of tests it ran through
    diagnostics: [
        {
            fileName: 'add.test-d.ts',
            message: 'A message here',
            severity: 'error',
            line: 10,
            column: 34,
        },
        ...
    ]
}
```

This PR is still a "WIP". Documentation needs to be updated and new tests has to be written to make sure the new functionality is properly tested. 

---

- A more detailed explanation can be found here - [Reporting Test Count](https://smartclash.github.io/public-notes/report-tests-count)
- Fixes #12

### CODE REVIEW

**Review by SaurabhAgarwala — COMMENTED** [review]
On testing it locally, I discovered that the total number of tests being reported is incorrect. Tried looking into the problematic code but have a doubt. Discussed it over Discord and it has to be fixed after few clarifications.

File: source/lib/index.ts [hunk 1]
```diff
@@ -75,8 +78,15 @@ export default async (options: Options = {cwd: process.cwd()}) => {
 		config
 	};
 
-	return [
-		...getCustomDiagnostics(context),
-		...getTSDiagnostics(context)
-	];
+	const tsDiagnostics = getTSDiagnostics(context);
+	const customDiagnostics = getCustomDiagnostics(context);
+	const numTests = tsDiagnostics.numTests + customDiagnostics.numTests;
```
Comment (SaurabhAgarwala) [inline source/lib/index.ts:83]: Looks like the incorrect number of tests are coming from here. This is has to be further looked into.

Comment (SaurabhAgarwala) [inline source/lib/index.ts]: Looks like the incorrect number of tests are coming from here. This is has to be further looked into.

**Comment by smartclash** [comment]
Merging this to master. Let's publish it folks!

**Pull request merged**